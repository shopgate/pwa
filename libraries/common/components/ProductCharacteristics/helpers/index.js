/**
 * Resets any characteristics inside a selection that come after the given index.
 * @param {Array} characteristics The characteristics of a product.
 * @param {Object} selection The current selected characteristics.
 * @param {number} index The index to reset the selection to.
 * @return {Object}
 */
export function reduceSelection(characteristics, selection, index) {
  const currentSelection = {
    ...selection,
  };

  /**
   * When the given index is the same as the complete set
   * of characteristics then there is nothing to reset.
   */
  if (index === (characteristics.length - 1)) {
    return currentSelection;
  }

  // Find the selections after the given index.
  const after = characteristics.slice(index + 1);

  // Delete the found selections.
  after.forEach((item) => {
    delete currentSelection[item.id];
  });

  return currentSelection;
}

/**
 * Returns the index of a particular characteristic from a set of characteristics.
 * @param {Array} characteristics The characteristics of a product.
 * @param {number} characteristicId The id of the characteristic to find.
 * @return {number}
 */
export function findSelectionIndex(characteristics, characteristicId) {
  const target = characteristics.findIndex(char => char.id === characteristicId);

  if (!target) {
    return -1;
  }

  return target;
}

/**
 * Returns true if the characteristic at the given index is the next characteristic to be selected.
 * @param {Array} characteristics The characteristics of a product.
 * @param {number} index The index of the characteristic to check.
 * @return {boolean}
 */
export function isCharacteristicEnabled(characteristics, index) {
  if (index === 0) {
    return true;
  }

  return !!Object.values(characteristics)[index - 1];
}

/**
 * Returns the selected value by characteristic ID.
 * @param {string} charId The characteristic ID.
 * @param {Object} characteristics The already selected characteristics.
 * @return {string|null}
 */
export function getSelectedValue(charId, characteristics) {
  return characteristics[charId] ? characteristics[charId] : null;
}

/**
 * Prepares the new state after a selection has been made.
 * @param {string} id The selection ID
 * @param {Object} selections The selections stored in the state.
 * @param {Array} characteristics The characteristics of a product.
 * @return {Object}
 */
export function prepareState(id, selections, characteristics) {
  if (!selections[id]) {
    return selections;
  }

  // Find the current index.
  const currentIndex = findSelectionIndex(characteristics, id);

  return reduceSelection(characteristics, selections, currentIndex);
}
