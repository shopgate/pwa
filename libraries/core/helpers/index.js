import { emitter } from '../classes/ErrorManager';
import { SOURCE_CONSOLE } from '../constants/ErrorManager';

const csl = console;

/**
 * Converts logs into a readable format for Selenium, if testMode is enabled.
 * @param {Array} args Arguments that should be converted
 * @return {Array} converted arguments
 */
const convertLogArgs = (args) => {
  if (window.stringifyLogs) {
    return [JSON.stringify(args)];
  }

  return args;
};

/**
 * The logging wrapper for the console.
 * @type {Object}
 */
export const logger = {
  ...console,
  debug: (...args) => csl.debug(...convertLogArgs(args)),
  dir: (...args) => csl.dir(...convertLogArgs(args)),
  dirxml: (...args) => csl.dirxml(...convertLogArgs(args)),
  error: (...args) => {
    emitter.emit(SOURCE_CONSOLE, args);
    csl.error(...convertLogArgs(args));
  },
  info: (...args) => csl.info(...convertLogArgs(args)),
  log: (...args) => csl.log(...convertLogArgs(args)),
  warn: (...args) => csl.warn(...convertLogArgs(args)),
  assert: (...args) => csl.assert(...convertLogArgs(args)),
};

/**
 * Returns a URL for performing XHR Requests.
 * @param {string} action The action to request on the server.
 * @return {string} The full URL.
 */
export const ajaxUrl = action => (action ? `sgapi:${action}` : '');

/**
 * Checks if the hasSGJavascriptBridge exists.
 * @return {boolean}
 */
export function hasSGJavaScriptBridge() {
  return (typeof SGJavascriptBridge !== 'undefined');
}

/**
 * Checks whether the web bridge is active.
 * @returns {boolean}
 */
export function hasWebBridgeCore() {
  if (!hasSGJavaScriptBridge()) {
    return false;
  }
  return window.SGJavascriptBridge?.type === 'web' || window.SGJavascriptBridge?.type === 'desktop';
}

/**
 * Checks whether the browser connector should be used.
 * @return {boolean}
 */
export function useBrowserConnector() {
  if (hasSGJavaScriptBridge()) {
    return false;
  }

  if (process.env.NODE_ENV === 'development' && process.env.IP && process.env.PORT) {
    return false;
  }

  return true;
}

/**
 * Logs a deprecation warning.
 * @param {string} element The deprecated element.
 * @param {string} [version='v7.0.0'] The engage version of removal.
 */
export function logDeprecationMessage(element, version = 'v7.0.0') {
  logger.warn(`DEPRECATED: ${element} is deprecated. It will be removed in Engage ${version}`);
}
