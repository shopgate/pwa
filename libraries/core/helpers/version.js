import MobileDetect from 'mobile-detect';
import detector from 'detector';
import { logger, useBrowserConnector } from './index';
import { getWebStorageEntry } from '../commands/webStorage';

const MODE_AT_MOST = 'at_most';
const MODE_AT_LEAST = 'at_least';

export const PLATFORM_ANDROID = 'android';
export const PLATFORM_IOS = 'ios';
export const MIN_ANDROID_LIB_VERSION = '13.0';

let versions = null;
let requesting = false;

const md = new MobileDetect(navigator.userAgent);
const isAndroid = md.is('AndroidOS');
const fullVersion = typeof detector === 'object' && detector.os ? detector.os.fullVersion : null;

// Eslint doesn't allow to use one liner here.
let model = null;
if (fullVersion) {
  model = isAndroid ? 'Android' : `iPhone${fullVersion}`;
}

export const defaultClientInformation = {
  libVersion: '17.0',
  appVersion: '5.18.0',
  codebaseVersion: '5.18.0',
  type: (!md.tablet() ? 'phone' : 'tablet'),
  device: {
    type: (!md.tablet() ? 'phone' : 'tablet'),
    os: {
      platform: isAndroid ? PLATFORM_ANDROID : PLATFORM_IOS,
      ver: fullVersion,
    },
    model,
  },
};

/**
 * Waits till a passed handler function returns TRUE and resolves.
 * @param {Function} handler Handler function which checks a condition to resolve.
 * @return {Promise}
 */
const wait = handler => new Promise((resolve) => {
  /**
   * Checks the request state.
   */
  const check = () => {
    if (handler() === true) {
      resolve();
    } else {
      setTimeout(check, 20);
    }
  };
  check();
});

/**
 * Splits a version string into it's segments.
 * @param {string} input The string to parse.
 * @return {Array}
*/
const parseVersionSegments = (input) => {
  if (typeof input !== 'string') {
    return [Number.NaN];
  }

  return input.trim().split('.').map(segment => parseInt(segment, 10));
};

/**
 * Validates version segments.
 * @param {Array} segments The segments array.
 * @return {boolean}
 */
const validateVersionSegments = segments =>
  segments.filter(entry => Number.isNaN(entry)).length === 0;

/**
 * Checks if a required version matches a current version.
 * @param {string} mode The mode for the check.
 * @param {string} requiredVersion The required version - 17[|17.0|17.0.0].
 * @param {string} currentVersion The current version (same pattern).
 * @return {boolean}
 */
const check = (mode, requiredVersion = '', currentVersion = '') => {
  const parsedCurrent = parseVersionSegments(currentVersion);
  const parsedRequired = parseVersionSegments(requiredVersion);

  if (!validateVersionSegments(parsedCurrent) || !validateVersionSegments(parsedRequired)) {
    logger.error(`Error parsing version strings (current: ${currentVersion} | required: ${requiredVersion})`);
    return false;
  }

  const [requiredMajor, requiredMinor = 0, requiredMicro = 0] = parsedRequired;
  const [currentMajor, currentMinor = 0, currentMicro = 0] = parsedCurrent;

  if (mode === MODE_AT_LEAST) {
    if (currentMajor !== requiredMajor) {
      return (currentMajor > requiredMajor);
    }

    if (currentMinor !== requiredMinor) {
      return (currentMinor > requiredMinor);
    }

    return (currentMicro >= requiredMicro);
  }

  // MODE_AT_MOST - no extra branch necessary, since it's just an internal function right now.
  if (currentMajor !== requiredMajor) {
    return (currentMajor < requiredMajor);
  }

  if (currentMinor !== requiredMinor) {
    return (currentMinor < requiredMinor);
  }

  return (currentMicro <= requiredMicro);
};

/**
 * Checks if a version string is valid.
 * @param {string} input The string to check - 17[|17.0|17.0.0].
 * @return {boolean}
 */
export const isValidVersion = input =>
  validateVersionSegments(parseVersionSegments(input));

/**
 * Checks if a required version matches at least a current version.
 * @param {string} requiredVersion The required version - 17[|17.0|17.0.0].
 * @param {string} currentVersion The current version (same pattern).
 * @return {boolean}
 */
export const isVersionAtLeast = (requiredVersion, currentVersion) =>
  check(MODE_AT_LEAST, requiredVersion, currentVersion);

/**
 * Checks if a required version matches at most a current version.
 * @param {string} requiredVersion The required version - 17[|17.0|17.0.0].
 * @param {string} currentVersion The current (same pattern).
 * @return {boolean}
 */
export const isVersionAtMost = (requiredVersion, currentVersion) =>
  check(MODE_AT_MOST, requiredVersion, currentVersion);

/**
 * Checks if a required version is equal to a current version.
 * @param {string} requiredVersion The required version - 17[|17.0|17.0.0].
 * @param {string} currentVersion The current (same pattern).
 * @return {boolean}
 */
export const isVersion = (requiredVersion, currentVersion) =>
  (isVersionAtLeast(requiredVersion, currentVersion) === true) &&
  (isVersionAtMost(requiredVersion, currentVersion) === true);

/**
* Fetches versions from the client information webStorage entry.
* @return {Promise}
*/
const getVersionsFromClientInformation = async () => {
  /**
   * Checks if currently no request is ongoing.
   * @return {boolean}
   */
  const isIdle = () => requesting === false;
  if (!isIdle()) {
    // Wait until the request finished before proceed.
    await wait(isIdle);
  }
  // If a version object is already present, the request can be skipped.
  if (versions !== null) {
    return versions;
  }
  requesting = true;
  let clientInformation;

  // Fetch the client information.
  if (useBrowserConnector()) {
    clientInformation = { value: { ...defaultClientInformation } };
  } else {
    clientInformation = await getWebStorageEntry({ name: 'clientInformation' });
  }

  // Grab all relevant data from the client information.
  let { libVersion } = clientInformation.value;
  const {
    appVersion,
    codebaseVersion,
    device,
  } = clientInformation.value;
  /* istanbul ignore next */
  const { os: { platform = null } = {} } = device || {};
  /**
  * Older Android app versions didn't handle the libVersion within the client information like the
  * iOS apps. Those apps usually send a version 2.0. But since they also support most of the PWA
  * related commands, the lib version is corrected here to improve handling within the code.
  */
  if (platform === PLATFORM_ANDROID && !isVersionAtLeast('9.0', libVersion)) {
    libVersion = MIN_ANDROID_LIB_VERSION;
  }
  // Update the version cache.
  versions = {
    libVersion,
    appVersion,
    codebaseVersion,
  };
  requesting = false;
  return versions;
};

/**
 * Clears the local versions cache
 */
export const clearVersionCache = () => {
  versions = null;
};

/**
 * Checks if a required version matches at least the current lib version.
 * @param {string} requiredVersion The required version - 17[|17.0|17.0.0].
 * @return {Promise}
 */
export const isLibVersionAtLeast = async (requiredVersion) => {
  const { libVersion } = await getVersionsFromClientInformation();
  return isVersionAtLeast(requiredVersion, libVersion);
};

/**
 * Checks if a required version matches at most the current lib version.
 * @param {string} requiredVersion The required version - 17[|17.0|17.0.0].
 * @return {Promise}
 */
export const isLibVersionAtMost = async (requiredVersion) => {
  const { libVersion } = await getVersionsFromClientInformation();
  return isVersionAtMost(requiredVersion, libVersion);
};

/**
 * Checks if a required version matches the current lib version.
 * @param {string} requiredVersion The required version - 17[|17.0|17.0.0].
 * @return {Promise}
 */
export const isLibVersion = async (requiredVersion) => {
  const { libVersion } = await getVersionsFromClientInformation();
  return isVersion(requiredVersion, libVersion);
};

/**
 * Returns the current libVersion.
 * @return {Promise}
 */
export const getLibVersion = async () => {
  const { libVersion } = await getVersionsFromClientInformation();
  return libVersion;
};
