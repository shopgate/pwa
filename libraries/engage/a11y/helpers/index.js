export * from '../components/LiveMessenger/helpers';
