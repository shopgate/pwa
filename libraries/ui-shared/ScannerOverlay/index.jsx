import React, {
  Fragment, useState, useEffect, memo,
} from 'react';
import PropTypes from 'prop-types';
import AppScanner from '@shopgate/pwa-core/classes/Scanner';
import CameraOverlay from './components/CameraOverlay';
import ScannerBar from './components/ScannerBar';

/**
 * The scanner overlay component.
 * @param {Object} props The component props.
 * @returns {JSX}
 */
function ScannerOverlay(props) {
  const [flashlight, setFlashlight] = useState(props.flashlight);

  useEffect(() => {
    setFlashlight(props.flashlight);
  }, [props.flashlight]);

  /**
   * Handles toggling the flashlight.
   */
  function handleToggleFlashlight() {
    setFlashlight(AppScanner.toggleFlashlight());
  }

  return (
    <Fragment>
      <CameraOverlay />
      <ScannerBar flashlightState={flashlight} onToggleFlashlight={handleToggleFlashlight} />
    </Fragment>
  );
}

ScannerOverlay.propTypes = {
  flashlight: PropTypes.bool,
};

ScannerOverlay.defaultProps = {
  flashlight: false,
};

export default memo(ScannerOverlay);
