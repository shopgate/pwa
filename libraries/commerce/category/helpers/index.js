import { bin2hex } from '@shopgate/pwa-common/helpers/data';
import { i18n } from '@shopgate/engage/core';
import { CATEGORY_PATH } from '../constants';

/**
 * Generate category route for navigation.
 * @param {string} id category Id
 * @returns {string}
 */
export const getCategoryRoute = id => `${CATEGORY_PATH}/${bin2hex(id)}`;

/**
 * Generate filters for show all products.
 * @param {Object} parentCategory parent category
 * @returns {Object}
 */
export const getShowAllProductsFilters = (parentCategory) => {
  const filters = {
    categories: {
      id: 'categories',
      label: i18n.text('filters.label.category'),
      source: 'categories',
      type: 'multiselect',
      isHidden: true,
      value: [{
        id: parentCategory ? parentCategory.path : null,
        label: parentCategory ? parentCategory.name : null,
        isHidden: true,
      }],
    },
  };

  return filters;
};
