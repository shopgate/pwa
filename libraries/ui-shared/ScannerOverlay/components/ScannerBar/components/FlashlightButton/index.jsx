import React from 'react';
import PropTypes from 'prop-types';
import ToggleIcon from '@shopgate/pwa-ui-shared/ToggleIcon';
import FlashEnabledIcon from '@shopgate/pwa-ui-shared/icons/FlashEnabledIcon';
import FlashDisabledIcon from '@shopgate/pwa-ui-shared/icons/FlashDisabledIcon';
import { SCANNER_FLASH } from '@shopgate/pwa-common-commerce/scanner/constants/Portals';
import { SurroundPortals } from '@shopgate/engage/components';
import { i18n } from '@shopgate/engage/core';
import styles from './style';

/**
 * The FlashlightButton component.
 * @returns {JSX.Element}
 */
const FlashlightButton = ({
  flashlightState,
  onToggle,
}) => (
  <SurroundPortals portalName={SCANNER_FLASH}>
    <button
      className={styles.button}
      onClick={onToggle}
      role="link"
      type="button"
      aria-label={i18n.text(flashlightState ? 'scanner.flashlight.switchOff' : 'scanner.flashlight.switchOn')}
    >
      <ToggleIcon
        on={flashlightState}
        onIcon={<FlashEnabledIcon className={styles.icon} />}
        offIcon={<FlashDisabledIcon className={styles.icon} />}
      />
    </button>
    <div className="sr-only" role="status" aria-live="polite">
      {i18n.text(flashlightState ? 'scanner.flashlight.on' : 'scanner.flashlight.off')}
    </div>
  </SurroundPortals>
);

FlashlightButton.propTypes = {
  flashlightState: PropTypes.bool.isRequired,
  onToggle: PropTypes.func.isRequired,
};

export default FlashlightButton;
