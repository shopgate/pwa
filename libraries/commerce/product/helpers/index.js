import appConfig from '@shopgate/pwa-common/helpers/config';
import { generateResultHash } from '@shopgate/pwa-common/helpers/redux';
import { bin2hex, hex2bin } from '@shopgate/pwa-common/helpers/data';
import {
  ITEM_PATH,
  PROPERTIES_FILTER_BLACKLIST,
  PROPERTIES_FILTER_WHITELIST,
  PRODUCT_RELATIONS_DEFAULT_LIMIT,
} from '../constants/';
import { SHOPGATE_CATALOG_GET_PRODUCT_RELATIONS } from '../constants/Pipelines';

/**
 * Reads the setting for product properties whitelisting/blacklisting
 * and filters the properties accordingly.
 * @param {Array} properties Array of properties with label and value attributes.
 * @return {Array} Filtered properties.
 */
export const filterProperties = (properties) => {
  // If there is no setting, we won't filter
  if (!appConfig || !appConfig.productPropertiesFilter) {
    return properties;
  }

  return properties.filter(({ label }) => {
    const inList = appConfig.productPropertiesFilter.properties.indexOf(label) !== -1;

    if (appConfig.productPropertiesFilter.type === PROPERTIES_FILTER_WHITELIST) {
      // Show item if allowed in whitelist
      return inList;
    }

    if (appConfig.productPropertiesFilter.type === PROPERTIES_FILTER_BLACKLIST) {
      // Hide item if not allowed in blacklist
      return !inList;
    }

    return true;
  });
};

/**
 * Generates getProductRelations hash based on productId, type and given limit.
 * It is used in both action and selectors.
 * @param {string} productId Product Id.
 * @param {string} type Type - see `../constants`.
 * @param {number} limit Limit.
 * @returns {string}
 */
export const generateProductRelationsHash = ({
  productId,
  type,
  limit = PRODUCT_RELATIONS_DEFAULT_LIMIT,
}) => generateResultHash({
  pipeline: SHOPGATE_CATALOG_GET_PRODUCT_RELATIONS,
  productId,
  type,
  limit,
}, false, false);

/**
 * Generate product route for navigation.
 * @param {string} id Product Id.
 * @returns {string}
 */
export const getProductRoute = id => `${ITEM_PATH}/${bin2hex(id)}`;

/**
 * Transform product route.
 * @param {Object} route item route to transform.
 * @returns {Object}
 */
export const transformRoute = (route) => {
  if (!route.params.productId || route.state.productId) {
    return route;
  }

  return {
    ...route,
    state: {
      ...route.state,
      productId: hex2bin(route.params.productId),
    },
  };
};

