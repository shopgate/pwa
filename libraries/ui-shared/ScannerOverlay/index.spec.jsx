import React from 'react';
import { mount } from 'enzyme';
import { defaultClientInformation } from '@shopgate/pwa-core/helpers/version';
import AppScanner from '@shopgate/pwa-core/classes/Scanner';
import {
  SCANNER_SCOPE_DEFAULT,
  SCANNER_TYPE_BARCODE,
  SCANNER_MIN_APP_LIB_VERSION,
} from '@shopgate/pwa-core/constants/Scanner';
import CameraOverlay from './components/CameraOverlay';
import FlashlightButton from './components/ScannerBar/components/FlashlightButton';
import ScannerOverlay from './index';

jest.mock('@shopgate/pwa-core/classes/AppCommand');
jest.mock('react-dom', () => ({
  createPortal: element => (
    <div id="portal-content">{element}</div>
  ),
}));

jest.mock('@shopgate/engage/components');

defaultClientInformation.libVersion = SCANNER_MIN_APP_LIB_VERSION;

describe('<ScannerOverlay />', () => {
  const toggleFlashlightSpy = jest.spyOn(AppScanner, 'toggleFlashlight');

  it('should render the component', () => {
    const wrapper = mount(<ScannerOverlay />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(CameraOverlay)).not.toBeEmptyRender();
    expect(wrapper.find(CameraOverlay)).not.toBeEmptyRender();
    expect(wrapper.state('flashlight')).toBe(false);
  });

  it('should toggle the flashlight when the button is pressed', async () => {
    await AppScanner.open(SCANNER_SCOPE_DEFAULT, SCANNER_TYPE_BARCODE);

    const wrapper = mount(<ScannerOverlay />);
    const button = wrapper.find(FlashlightButton).find('ToggleIcon');

    button.simulate('click');

    expect(wrapper.state('flashlight')).toBe(true);
    expect(toggleFlashlightSpy).toHaveBeenCalledTimes(1);

    button.simulate('click');
    expect(wrapper.state('flashlight')).toBe(false);
    expect(toggleFlashlightSpy).toHaveBeenCalledTimes(2);
  });
});

