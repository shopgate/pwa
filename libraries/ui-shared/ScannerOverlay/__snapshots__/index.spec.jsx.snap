// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScannerOverlay /> should render the component 1`] = `
<ScannerOverlay
  flashlight={false}
>
  <CameraOverlay>
    <SurroundPortals
      portalName="scanner.camera"
      portalProps={null}
    >
      <Portal
        name="scanner.camera.before"
        props={null}
      />
      <Portal
        name="scanner.camera"
        props={null}
      >
        <div
          className="css-njxfh"
        >
          <div />
        </div>
      </Portal>
      <Portal
        name="scanner.camera.after"
        props={null}
      />
    </SurroundPortals>
  </CameraOverlay>
  <ScannerBar
    flashlightState={false}
    onToggleFlashlight={[Function]}
  >
    <div
      id="portal-content"
    >
      <SurroundPortals
        portalName="scanner.bar"
        portalProps={
          Object {
            "flashlightState": false,
            "onToggleFlashlight": [Function],
          }
        }
      >
        <Portal
          name="scanner.bar.before"
          props={
            Object {
              "flashlightState": false,
              "onToggleFlashlight": [Function],
            }
          }
        />
        <Portal
          name="scanner.bar"
          props={
            Object {
              "flashlightState": false,
              "onToggleFlashlight": [Function],
            }
          }
        >
          <Grid
            className="css-1je6emb ui-shared__scanner-bar"
            component="ul"
            wrap={false}
          >
            <ul
              className="css-1je6emb ui-shared__scanner-bar css-h7r55a"
            >
              <GridItem
                className="css-1uot7vp"
                component="li"
                grow={0}
                shrink={1}
              >
                <li
                  className="css-1uot7vp"
                >
                  <FlashlightButton
                    flashlightState={false}
                    onToggle={[Function]}
                  >
                    <SurroundPortals
                      portalName="scanner.flash"
                      portalProps={null}
                    >
                      <Portal
                        name="scanner.flash.before"
                        props={null}
                      />
                      <Portal
                        name="scanner.flash"
                        props={null}
                      >
                        <button
                          className="css-1yh9ja6"
                          onClick={[Function]}
                          role="link"
                          type="button"
                        >
                          <ToggleIcon
                            offIcon={
                              <FlashDisabled
                                className="css-2rq5n9"
                              />
                            }
                            on={false}
                            onIcon={
                              <FlashEnabled
                                className="css-2rq5n9"
                              />
                            }
                            toggleHandler={[Function]}
                          >
                            <div
                              aria-hidden={true}
                              className="ui-shared__toggle-icon"
                              onClick={[Function]}
                            >
                              <FlashDisabled
                                className="css-2rq5n9"
                              >
                                <Icon
                                  className="css-2rq5n9"
                                  color={null}
                                  content="<path d=\\"M15.22,3.1,12,9.51l-.8,1.59h3.86l-4,6.83v-6h-3V3.1h7.12M17,2h0Zm0,0H7V13h3v9l7-12H13l4-8Z\\" />"
                                  size="inherit"
                                  viewBox="0 0 24 24"
                                >
                                  <svg
                                    className="css-yys9hb css-2rq5n9"
                                    dangerouslySetInnerHTML={
                                      Object {
                                        "__html": "<path d=\\"M15.22,3.1,12,9.51l-.8,1.59h3.86l-4,6.83v-6h-3V3.1h7.12M17,2h0Zm0,0H7V13h3v9l7-12H13l4-8Z\\" />",
                                      }
                                    }
                                    style={
                                      Object {
                                        "fill": null,
                                        "fontSize": "inherit",
                                      }
                                    }
                                    viewBox="0 0 24 24"
                                    xmlns="http://www.w3.org/2000/svg"
                                  />
                                </Icon>
                              </FlashDisabled>
                            </div>
                          </ToggleIcon>
                        </button>
                      </Portal>
                      <Portal
                        name="scanner.flash.after"
                        props={null}
                      />
                    </SurroundPortals>
                  </FlashlightButton>
                </li>
              </GridItem>
              <GridItem
                className="css-1uot7vp"
                component="li"
                grow={0}
                shrink={1}
              >
                <li
                  className="css-1uot7vp"
                >
                  <ScannerInstructions>
                    <SurroundPortals
                      portalName="scanner.instructions"
                      portalProps={null}
                    >
                      <Portal
                        name="scanner.instructions.before"
                        props={null}
                      />
                      <Portal
                        name="scanner.instructions"
                        props={null}
                      >
                        <Translate
                          className={null}
                          params={Object {}}
                          role={null}
                          string="scanner.instructions"
                          transform={null}
                        >
                          <span
                            className={null}
                            role={null}
                          >
                            scanner.instructions
                          </span>
                        </Translate>
                      </Portal>
                      <Portal
                        name="scanner.instructions.after"
                        props={null}
                      />
                    </SurroundPortals>
                  </ScannerInstructions>
                </li>
              </GridItem>
            </ul>
          </Grid>
        </Portal>
        <Portal
          name="scanner.bar.after"
          props={
            Object {
              "flashlightState": false,
              "onToggleFlashlight": [Function],
            }
          }
        />
      </SurroundPortals>
    </div>
  </ScannerBar>
</ScannerOverlay>
`;
